File: learning/01-pattern/00-backtracking/00-eight-queens-recursion.py
# https://o2.edu.vn/python-bai-toan-xep-hau-su-dung-de-quy/

# backtracking solution

n = 8
a = n * [0]
# a = [0, 4, 7, 5, 2, 6, 1, 3]


def print_board(b):
    l = len(b)
    for i in range(l):
        for j in range(l):
            if j == a[i]:
                print(1, end=" ")
            else:
                print(0, end=" ")
        print()


def possible(x, y):
    for i in range(x):
        if a[i] == y or x + y == i + a[i] or x - y == i - a[i]:
            return False
    return True


# Recursion function
count = 0


def gen(i, n):
    global count
    for j in range(n):
        if possible(i, j):
            a[i] = j
            if i == n - 1:
                count += 1
                # print(a)
            gen(i + 1, n)


# print_board(a)
gen(0, n)
print(count)

File: learning/01-pattern/00-recursion-backtracking/00-eight-queens-recursion.js
// 8 queens puzzles

const size = 8;
let arr = Array.from({ length: size }, () => 0);
// a = [0, 4, 7, 5, 2, 6, 1, 3];

function printBoard(a) {
  for (let i = 0; i < size; ++i) {
    str = '';
    for (let j = 0; j < n; ++j) {
      if (a[i] == j) str += '1 ';
      else str += '0 ';
    }
    console.log(str);
  }
}

// printBoard(a);
function possible(x, y) {
  for (let i = 0; i < x; ++i) {
    if (arr[i] == y || x + y == i + arr[i] || y - x == arr[i] - i) return false;
  }
  return true;
}

let count = 0;

function gen(i, n) {
  for (let j = 0; j < n; ++j) {
    if (possible(i, j)) {
      arr[i] = j;
      if (i == n - 1) {
        count += 1;
        console.log(arr);
      }

      gen(i + 1, n);
    }
  }
}

gen(0, size);
console.log(count);
